### PR TITLE
Only access IOAPIC registers that are actually present

### PR DESCRIPTION
**Pull request by wom-bat**
IOAPICS can have varying numbers of lines attached.
The actual number can be accessed in the top 16 bits of the
version register.

Rather than assuming fixed 24 lines per IRQ, read the actual number
and use that.

Signed-off-by: Peter Chubb <peter.chubb@unsw.edu.au>

### CODE REVIEW

File: src/plat/pc99/machine/ioapic.c [hunk 1]
```diff
@@ -118,8 +134,8 @@ exception_t ioapic_decode_map_pin_to_vector(word_t ioapic, word_t pin, word_t le
         current_syscall_error.rangeErrorMax = num_ioapics - 1;
         return EXCEPTION_SYSCALL_ERROR;
     }
-    if (pin >= IOAPIC_IRQ_LINES) {
-        userError("Invalid IOAPIC pin %ld, there are %d pins", (long)pin, IOAPIC_IRQ_LINES);
+    if (pin >= ioapic_nirqs[ioapic]) {
+        userError("Invalid IOAPIC pin %ld, there are %d pins", (long)pin, ioapic_nirqs[ioapic]);
         current_syscall_error.type = seL4_RangeError;
         current_syscall_error.rangeErrorMin = 0;
         current_syscall_error.rangeErrorMax = IOAPIC_IRQ_LINES - 1;
```
Comment (kent-mcleod) [inline src/plat/pc99/machine/ioapic.c:141]: Should this rangeErrorMax return value also be updated?

**Review by kent-mcleod — APPROVED** [review]
This improvement seems reasonable to me.  Maybe a comment should be added to `ioredtbl_high` to explain that `IOAPIC_IRQ_LINES` is used as an upper bound of how much memory to reserve for book keeping.

File: src/plat/pc99/machine/ioapic.c (hunk 1 above)
Comment (wom-bat) [reply src/plat/pc99/machine/ioapic.c:141]: I think you're right.  Will adjust.

File: src/plat/pc99/machine/ioapic.c [hunk 2]
```diff
@@ -52,12 +57,23 @@ static uint32_t ioapic_read(uint32_t ioapic, word_t reg)
...
+
+    ioapic_write(ioapic, IOAPIC_REGSEL, IOAPIC_REG_IOAPICVER);
+    nirqs = (ioapic_read(ioapic, IOAPIC_WINDOW) >> 16) + 1;
+    ioapic_nirqs[ioapic] = nirqs;
+
+    /*
+     * All current implementations have 24 or fewer lines
+     * Protect against the future one that may have more
+     */
+    assert(nirqs <= IOAPIC_IRQ_LINES);
```
Comment (lsf37) [inline src/plat/pc99/machine/ioapic.c:70]: `assert` will not be running in release configs, and verification does need to know that all values in that array are bounded. It looks like this is kernel init code, so if there is a condition where the hardware does not meet expectations, it should properly abort initialisation and bail out, not assert.

**Review by lsf37 — COMMENTED** [review]
This will need a small proof + spec update, but looks doable with reasonable effort. @mbrcknl might remember more about those x64 proofs than I.

File: src/plat/pc99/machine/ioapic.c (hunk 2 above)
Comment (wom-bat) [reply src/plat/pc99/machine/ioapic.c:70]: OK, I'll do something different.  I think it's important at this point that we do bail out.

Comment (wom-bat) [reply src/plat/pc99/machine/ioapic.c]: OK I think this is fixed now.

File: src/plat/pc99/machine/ioapic.c (hunk 1 above)
Comment (wom-bat) [reply src/plat/pc99/machine/ioapic.c]: Done.

**Comment by wom-bat** [comment]
> Maybe a comment should be added to ioredtbl_high to explain that IOAPIC_IRQ_LINES is used as an upper bound of how much memory to reserve for book keeping.


Fixed.

**Comment by kent-mcleod** [comment]
Is this blocked on a verification spec update?

**Comment by lsf37** [comment]
> Is this blocked on a verification spec update?

Spec and proof, yes.

**Comment by wom-bat** [comment]
Problem I'm trying to solve is machine hang when attempting to access non-existent registers.

**Comment by kent-mcleod** [comment]
> Spec and proof, yes.

@lsf37, are proof updates scheduled or in progres?

> Problem I'm trying to solve is machine hang when attempting to access non-existent registers.

@wom-bat, is this urgent? I don't think we can merge it until proofs are ready. If proof updates aren't scheduled and the changes are urgent, we could put it behind a temporary compile-time config? Otherwise the PR is blocked until proofs are done AFAIK.

**Comment by lsf37** [comment]
> > Spec and proof, yes.
> 
> @lsf37, are proof updates scheduled or in progres?

Unless someone has funding for it it'll have to wait until I have spare time (with a bit of luck in the week after next) or someone else picks it up.

File: src/plat/pc99/machine/ioapic.c [hunk 3]
```diff
@@ -118,11 +143,11 @@ exception_t ioapic_decode_map_pin_to_vector(word_t ioapic, word_t pin, word_t le
...
         return EXCEPTION_SYSCALL_ERROR;
     }
-    if (pin >= IOAPIC_IRQ_LINES) {
-        userError("Invalid IOAPIC pin %ld, there are %d pins", (long)pin, IOAPIC_IRQ_LINES);
+    if (pin >= ioapic_nirqs[ioapic]) {
+        userError("Invalid IOAPIC pin %ld, there are %d pins", (long)pin, ioapic_nirqs[ioapic]);
         current_syscall_error.type = seL4_RangeError;
         current_syscall_error.rangeErrorMin = 0;
-        current_syscall_error.rangeErrorMax = IOAPIC_IRQ_LINES - 1;
+        current_syscall_error.rangeErrorMax = ioapic_nirqs[ioapic] + 1;
```
Comment (lsf37) [inline src/plat/pc99/machine/ioapic.c:150]: Should this still be `- 1`?

Comment (Indanz) [reply src/plat/pc99/machine/ioapic.c:150]: Looks like it, yes.

**Comment by lsf37** [comment]
Rebased and squashed in preparation for the proof update.

**Review by lsf37 — APPROVED** [review]
Verification is done, this is now ready for merge.

**Pull request merged**